src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1493,19 +1493,24 @@
 
 ;; Wrap the fact-type so that Clojure equality and hashcode semantics are used
 ;; even though this is placed in a Java map.
-(deftype AlphaRootsWrapper [roots roots-hash]
+(deftype AlphaRootsWrapper [fact-type fact-type-hash roots]
   Object
   (equals [this other]
-    ;; Since we memoize the function that retrieves these on the fact type, and since the alpha roots for
-    ;; any given type primary type T (excluding descendants and ancestors) are a disjoint
-    ;; set with respect to the alpha roots for any other primary type T2, we know that we will
-    ;; only have one AlphaRootsWrapper object for any given set of alpha roots per get-alphas-fn instance.
-    ;; Therefore we can just use reference equality in the interest of maximizing performance.
-    (identical? this other))
+    (let [other ^AlphaRootsWrapper other]
+      (cond
+        
+        (identical? fact-type (.fact-type other)) 
+        true
+        
+        (not (== fact-type-hash (.fact-type-hash other)))
```

Comment:
I agree that it is possible to improve the performance by using primitives and ensuring that the call to == is inlined.  My concern is that we'd need to foresee that we wouldn't have any problems from inadvertent boxing that would cause errors.  This is certainly doable, but the performance numbers I'm seeing don't show any noticeable impact from == (it literally doesn't show up in VisualVM for the use case I'm profiling) and very little impact from the Java map operations inside the get-alphas-fn which encapsulate the == calls under discussion.  I'm inclined to go with the simple option for now, i.e. just use == without type hints, and revisit it later if we find a need.